front50-gcs/src/main/kotlin/com/netflix/spinnaker/front50/model/GcsStorageService.kt

```diff
@@ -0,0 +1,269 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.front50.model
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.cloud.storage.Blob
+import com.google.cloud.storage.BlobId
+import com.google.cloud.storage.BlobInfo
+import com.google.cloud.storage.BucketInfo
+import com.google.cloud.storage.Storage
+import com.google.cloud.storage.Storage.BlobListOption
+import com.google.cloud.storage.Storage.BucketField
+import com.google.cloud.storage.Storage.BucketGetOption
+import com.google.cloud.storage.StorageException
+import com.google.common.collect.ImmutableList
+import com.google.common.collect.ImmutableMap
+import com.netflix.spinnaker.front50.exception.NotFoundException
+import java.io.IOException
+import java.util.Comparator
+import java.util.concurrent.Executor
+import java.util.stream.Collectors
+import org.slf4j.LoggerFactory
+
+class GcsStorageService(
+  private val storage: Storage,
+  private val bucketName: String,
+  private val bucketLocation: String,
+  private val basePath: String,
+  private val dataFilename: String,
+  private val objectMapper: ObjectMapper,
+  private val executor: Executor
+) : StorageService {
+
+  companion object {
+    private val log = LoggerFactory.getLogger(GcsStorageService::class.java)
+    private const val LAST_MODIFIED_FILENAME = "last-modified"
+  }
+
+  private val modTimeState = ObjectType.values().map { it to ModificationTimeState(it) }.toMap()
+
+  fun ensureBucketExists() {
+    val bucket = storage.get(bucketName)
+    if (bucket == null) {
+      val bucketInfo = BucketInfo.newBuilder(bucketName)
+        .setVersioningEnabled(true)
+      if (bucketLocation.isNotBlank()) {
+        bucketInfo.setLocation(bucketLocation)
+      }
+      storage.create(bucketInfo.build())
+    }
+  }
+
+  override fun supportsVersioning(): Boolean {
+    val bucket = storage.get(bucketName, BucketGetOption.fields(BucketField.VERSIONING))
+    return bucket.versioningEnabled() == true
+  }
+
+  override fun <T : Timestamped> loadObject(objectType: ObjectType, objectKey: String): T {
+    try {
+      val blobId = blobIdForKey(objectType, objectKey)
+      val blob = storage.get(blobId)
+        ?: throw NotFoundException("Couldn't retrieve $objectType $objectKey from GCS")
+      val bytes = storage.readAllBytes(blob.blobId)
+      val obj: T = parseObject(bytes, objectType, objectKey)
+      obj.lastModified = blob.updateTime
+      return obj
+    } catch (e: StorageException) {
+      throw GcsStorageServiceException("error loading $objectType $objectKey", e)
+    }
+  }
+
+  override fun deleteObject(objectType: ObjectType, objectKey: String) {
+    try {
+      if (storage.delete(blobIdForKey(objectType, objectKey))) {
+        writeLastModified(objectType)
+      }
+    } catch (e: Exception) {
```

Comment:
I don't feel 100% sure about this, but the reason in Java why I think you want to catch specific exceptions is that if (say) we only catch `StorageException` and then a newer version of `Storage#delete` adds `IOException` to its `throws` list, we'll get a compile-time error and be forced to make a decision about how to handle it.

But since Kotlin doesn't have checked exception, that benefit no longer applies. If `Storage#delete` throws `IOException`, it will just bubble up to the caller. I would feel weirder about this if we were doing something more involved... but since all we're doing is wrapping the exception and adding _more_ debugging information, it doesn't feel particularly risky to me.

...except that I did forget about `InterruptedException`, so I wrote a `wrapException` method to handle that.

(Definitely would be curious if you found this argument convincing... I'm trying to convince myself here, too...)

Reply:
That argument makes sense, I guess I forgot that Kotlin doesn't have checked exceptions.  It does make it somewhat annoying to handle `InterruptedException` because in general that will always need to be handled separately...I wonder if there's some general pattern for that in Kotlin.